test/vm/fuzzer/fuzzer.cpp

```diff
@@ -494,105 +500,375 @@ static bool toss(Engine &engine, double p)
+                contract = std::move(new_contract);
+                iteration_count = 0;
+            }
+        }
+    }
+
+    // Make sure the final shrunken contract still fails
+    MONAD_VM_ASSERT(!try_singleton_run(run_index, args, contract, contract_address, failed_message));
+
+    return Run{.iterations = {Iteration{
+        .contract = contract,
+        .contract_address = contract_address,
+        .precompile = std::nullopt,
+        .delegated_contract_address = std::nullopt,
+        .messages = {failed_message}}}};
+}
+
+static Run shrink_run(std::size_t const run_index, arguments const &args, Run const &run, size_t failed_iteration_index, size_t failed_message_index)
+{
+    // Assume that the contract that failed didn't depend on any of the previous
+    // contracts. This is not always true, but most counter-examples are of this
+    // form.
+    // Prepare a run with only the msg.recipient contract and the message that
+    // caused the failure.
+    auto contract_map = std::unordered_map<evmc::address, std::vector<BasicBlock>>{};
+    for (auto const &it : run.iterations) {
+        contract_map.insert({it.contract_address, it.contract});
+    }
+
+    auto const &failed_iteration = run.iterations[failed_iteration_index];
+    auto const &failed_message = failed_iteration.messages[failed_message_index];
+    auto const &failed_contract_it = contract_map.find(failed_message.code_address);
+    if (failed_contract_it == contract_map.end()) {
+        std::cerr << "Failed to find contract that caused the failure\n";
+        return run;
+    }
+    auto const &failed_contract = failed_contract_it->second;
+
+    if (try_singleton_run(run_index, args, failed_contract, failed_message.code_address, failed_message)) {
+        std::cerr << "Shrinker: Contract depends on other contracts or state, cannot shrink.\n";
+        /*
+            To shrink the entire run, we'd need to find which of the previous
+            contracts/messages the failed contract execution depends on. A
+            solution could be to trace the execution of the failing run, and log
+            which accounts/storage slots were accessed during the execution of
+            the failing contract, retry with those contracts, and repeat until
+            we reach a fixed point before starting to shrink the contracts.
+
+            This is likely too complicated and not that effective comparing to
+            simply deleting contracts/messages from the run and seeing if the
+            failure still happens. To speed up the process, we can truncate
+            the tail of the run first (after the failing contract/message), then
+            reorder a few contracts and messages and try again.
+        */
+        return run;
+    } else {
+        return shrink_singleton_run(run_index, args, failed_contract, failed_message.code_address, failed_message);
```

Comment:
For now, only failed runs containing a single contract and call (i.e. the fuzzer failure still happens when running the contract and message in an empty environment) can be shrunk. All of the compiler "bugs" I introduced fit this criteria so I wasn't able to test the "full-run" shrinker and left it unimplemented.

The comment above describes how I would go and shrink full runs, I'm open to other ideas.